File: danser_autofetch/skins.py
"""
Skin Manager module.
Handles skin synchronization, direct .osk / .zip / URL importing, and fuzzy skin name matching.
"""

import os
import shutil
import zipfile
import urllib.request
from typing import List, Optional, Tuple


class SkinManager:
    def __init__(self, skins_dir: str, osu_exports_dir: Optional[str] = None):
        self.skins_dir = os.path.expanduser(skins_dir)
        os.makedirs(self.skins_dir, exist_ok=True)
        self.osu_exports_dir = os.path.expanduser(osu_exports_dir) if osu_exports_dir else None

    def import_skin(self, source_path_or_url: str) -> Optional[str]:
        """
        Imports a skin from a local file (.osk, .zip, directory) or a direct download URL.
        Returns the installed skin name if successful.
        """
        target_file = source_path_or_url

        # 1. Download if URL
        if source_path_or_url.startswith("http://") or source_path_or_url.startswith("https://"):
            print(f"📥 Downloading skin from: {source_path_or_url}...")
            temp_osk = os.path.join(self.skins_dir, "_temp_import.osk")
            try:
                req = urllib.request.Request(source_path_or_url, headers={"User-Agent": "danser-autofetch"})
                with urllib.request.urlopen(req, timeout=30) as in_f, open(temp_osk, "wb") as out_f:
                    out_f.write(in_f.read())
                target_file = temp_osk
            except Exception as e:
                print(f"❌ Failed to download skin from URL: {e}")
                return None

        target_file = os.path.expanduser(target_file)
        if not os.path.exists(target_file):
            return None

        # 2. If it's a directory
        if os.path.isdir(target_file):
            skin_name = os.path.basename(os.path.normpath(target_file))
            dst = os.path.join(self.skins_dir, skin_name)
            if os.path.abspath(target_file) != os.path.abspath(dst):
                if os.path.exists(dst):
                    shutil.rmtree(dst)
                shutil.copytree(target_file, dst)
            print(f"✅ Imported skin folder: '{skin_name}'")
            return skin_name

        # 3. If it's an archive (.osk / .zip)
        if target_file.endswith(".osk") or target_file.endswith(".zip"):
            raw_name = os.path.basename(target_file).rsplit(".", 1)[0]
            if raw_name == "_temp_import":
                raw_name = "Imported_Skin"
            # Clean up (1), (2) suffixes
            if " (" in raw_name and raw_name.endswith(")"):
                raw_name = raw_name.rsplit(" (", 1)[0]

            dst = os.path.join(self.skins_dir, raw_name)
            os.makedirs(dst, exist_ok=True)
            try:
                with zipfile.ZipFile(target_file, "r") as z:
                    z.extractall(dst)
                print(f"✅ Extracted & installed skin: '{raw_name}'")
                if target_file.endswith("_temp_import.osk"):
                    os.remove(target_file)
                return raw_name
            except Exception as e:
                print(f"❌ Failed to unpack skin archive: {e}")
                return None

        return None

    def sync_from_sources(self) -> int:
        """
        Scans osu! exports and Downloads directories for new .osk skins and imports them.
        """
        scan_dirs = []
        if self.osu_exports_dir and os.path.exists(self.osu_exports_dir):
            scan_dirs.append(self.osu_exports_dir)
        
        # Also check ~/Downloads for freshly downloaded .osk files
        downloads_dir = os.path.expanduser("~/Downloads")
        if os.path.exists(downloads_dir):
            scan_dirs.append(downloads_dir)

        count = 0
        for s_dir in scan_dirs:
            for item in os.listdir(s_dir):
                src = os.path.join(s_dir, item)
                if (item.endswith(".osk") or item.endswith(".zip")) and not item.startswith("."):
                    skin_name = item.rsplit(".", 1)[0]
                    if " (" in skin_name and skin_name.endswith(")"):
                        skin_name = skin_name.rsplit(" (", 1)[0]
                    dst = os.path.join(self.skins_dir, skin_name)
                    if not os.path.exists(dst):
                        if self.import_skin(src):
                            count += 1
                elif os.path.isdir(src) and s_dir == self.osu_exports_dir:
                    dst = os.path.join(self.skins_dir, item)
                    if not os.path.exists(dst):
                        shutil.copytree(src, dst)
                        count += 1
        return count

    def list_skins(self) -> List[str]:
        """Returns a list of all available skin names in Danser."""
        if not os.path.exists(self.skins_dir):
            return []
        return [
            d for d in os.listdir(self.skins_dir)
            if os.path.isdir(os.path.join(self.skins_dir, d))
        ]

    def match_skin(self, query: str) -> Optional[str]:
        """
        Matches a skin name or imports directly if query is a file path / URL.
        """
        # If query is a file path or URL, auto-import it on the fly!
        if (
            query.startswith("http://")
            or query.startswith("https://")
            or os.path.exists(os.path.expanduser(query))
        ):
            imported = self.import_skin(query)
            if imported:
                return imported

        available = self.list_skins()
        if not available:
            return None

        # 1. Exact match
        for s in available:
            if s.lower() == query.lower():
                return s

        # 2. Starts with query
        for s in available:
            if s.lower().startswith(query.lower()):
                return s

        # 3. Substring match
        for s in available:
            if query.lower() in s.lower():
                return s

        return None

File: src/skins.ts
/**
 * Skin Manager in TypeScript.
 * Handles on-demand skin importing from local .osk / .zip / folder paths, URLs, and fuzzy name matching.
 */

import * as fs from 'fs';
import * as path from 'path';
import AdmZip from 'adm-zip';
import { printStatus } from './ui';

export class SkinManager {
  private skinsDir: string;
  private osuExportsDir?: string;

  constructor(skinsDir: string, osuExportsDir?: string) {
    this.skinsDir = path.resolve(skinsDir.replace(/^~(?=$|\/|\\)/, process.env.HOME || process.env.USERPROFILE || ''));
    if (!fs.existsSync(this.skinsDir)) {
      fs.mkdirSync(this.skinsDir, { recursive: true });
    }
    if (osuExportsDir) {
      this.osuExportsDir = path.resolve(osuExportsDir.replace(/^~(?=$|\/|\\)/, process.env.HOME || process.env.USERPROFILE || ''));
    }
  }

  async importSkin(sourcePathOrUrl: string): Promise<string | null> {
    let rawPath = sourcePathOrUrl.trim().replace(/^["']|["']$/g, '');

    // 1. Direct URL download
    if (rawPath.startsWith('http://') || rawPath.startsWith('https://')) {
      printStatus('skin', `fetching skin from url: ${rawPath.toLowerCase()}...`);
      const tempOsk = path.join(this.skinsDir, '_temp_import.osk');
      try {
        const resp = await fetch(rawPath, { headers: { 'User-Agent': 'danser-autofetch' } });
        if (!resp.ok) throw new Error(`http error ${resp.status}`);
        const arrayBuffer = await resp.arrayBuffer();
        fs.writeFileSync(tempOsk, Buffer.from(arrayBuffer));
        rawPath = tempOsk;
      } catch (e: any) {
        printStatus('skin', `failed to download skin from url: ${e.message.toLowerCase()}`, 'error');
        return null;
      }
    }

    const cleanPath = path.resolve(rawPath.replace(/^~(?=$|\/|\\)/, process.env.HOME || process.env.USERPROFILE || ''));
    if (!fs.existsSync(cleanPath)) {
      return null;
    }

    const stat = fs.statSync(cleanPath);

    // 2. Local folder import
    if (stat.isDirectory()) {
      const skinName = path.basename(cleanPath);
      const dst = path.join(this.skinsDir, skinName);
      if (path.resolve(cleanPath) !== path.resolve(dst)) {
        if (fs.existsSync(dst)) {
          fs.rmSync(dst, { recursive: true, force: true });
        }
        fs.cpSync(cleanPath, dst, { recursive: true });
      }
      printStatus('skin', `imported skin folder: '${skinName.toLowerCase()}'`, 'success');
      return skinName;
    }

    // 3. Compressed skin archive (.osk / .zip)
    const lower = cleanPath.toLowerCase();
    if (lower.endsWith('.osk') || lower.endsWith('.zip')) {
      let rawName = path.basename(cleanPath).replace(/\.[^/.]+$/, '');
      if (rawName === '_temp_import') {
        rawName = 'imported_skin';
      }
      // Remove duplicate markers like (1), (2)
      if (rawName.includes(' (') && rawName.endsWith(')')) {
        rawName = rawName.replace(/\s*\(\d+\)$/, '').trim();
      }

      const dst = path.join(this.skinsDir, rawName);
      if (!fs.existsSync(dst)) {
        fs.mkdirSync(dst, { recursive: true });
      }

      try {
        const zip = new AdmZip(cleanPath);
        zip.extractAllTo(dst, true);
        printStatus('skin', `unpacked and installed skin: '${rawName.toLowerCase()}'`, 'success');
        if (cleanPath.endsWith('_temp_import.osk') && fs.existsSync(cleanPath)) {
          fs.unlinkSync(cleanPath);
        }
        return rawName;
      } catch (e: any) {
        printStatus('skin', `failed to unpack skin archive: ${e.message.toLowerCase()}`, 'error');
        return null;
      }
    }

    return null;
  }

  async syncFromSources(): Promise<number> {
    const scanDirs: string[] = [];
    if (this.osuExportsDir && fs.existsSync(this.osuExportsDir)) {
      scanDirs.push(this.osuExportsDir);
    }

    const downloadsDir = path.resolve((process.env.HOME || process.env.USERPROFILE || ''), 'Downloads');
    if (fs.existsSync(downloadsDir)) {
      scanDirs.push(downloadsDir);
    }

    let count = 0;
    for (const sDir of scanDirs) {
      const items = fs.readdirSync(sDir);
      for (const item of items) {
        const src = path.join(sDir, item);
        const lower = item.toLowerCase();
        if ((lower.endsWith('.osk') || lower.endsWith('.zip')) && !item.startsWith('.')) {
          let skinName = item.replace(/\.[^/.]+$/, '');
          if (skinName.includes(' (') && skinName.endsWith(')')) {
            skinName = skinName.replace(/\s*\(\d+\)$/, '').trim();
          }
          const dst = path.join(this.skinsDir, skinName);
          if (!fs.existsSync(dst)) {
            const imported = await this.importSkin(src);
            if (imported) count += 1;
          }
        } else if (fs.statSync(src).isDirectory() && sDir === this.osuExportsDir) {
          const dst = path.join(this.skinsDir, item);
          if (!fs.existsSync(dst)) {
            fs.cpSync(src, dst, { recursive: true });
            count += 1;
          }
        }
      }
    }
    return count;
  }

  listSkins(): string[] {
    if (!fs.existsSync(this.skinsDir)) {
      return [];
    }
    return fs.readdirSync(this.skinsDir).filter((d) => {
      const p = path.join(this.skinsDir, d);
      return fs.statSync(p).isDirectory();
    });
  }

  async matchSkin(query: string): Promise<string | null> {
    const cleanQuery = query.trim().replace(/^["']|["']$/g, '');

    // 1. Direct path check (as passed)
    const resolvedPath = path.resolve(cleanQuery.replace(/^~(?=$|\/|\\)/, process.env.HOME || process.env.USERPROFILE || ''));
    if (cleanQuery.startsWith('http://') || cleanQuery.startsWith('https://') || fs.existsSync(resolvedPath)) {
      const imported = await this.importSkin(cleanQuery);
      if (imported) return imported;
    }

    // 2. Check standard user directories if cleanQuery was passed as a filename
    const home = process.env.HOME || process.env.USERPROFILE || '';
    const candidateDirs = [
      path.join(home, 'Documents'),
      path.join(home, 'Downloads'),
      path.join(home, 'Desktop'),
    ];
    if (this.osuExportsDir) {
      candidateDirs.unshift(this.osuExportsDir);
    }
    for (const cDir of candidateDirs) {
      const candidateFile = path.join(cDir, cleanQuery);
      if (fs.existsSync(candidateFile)) {
        const imported = await this.importSkin(candidateFile);
        if (imported) return imported;
      }
    }

    // 3. Search already installed skins
    const available = this.listSkins();
    if (available.length === 0) return null;

    // Strip .osk / .zip extensions for name matching
    const queryBase = cleanQuery.replace(/\.(osk|zip)$/i, '').trim();

    // 3.1 Exact match
    for (const s of available) {
      if (s.toLowerCase() === queryBase.toLowerCase() || s.toLowerCase() === cleanQuery.toLowerCase()) {
        return s;
      }
    }

    // 3.2 Starts with
    for (const s of available) {
      if (s.toLowerCase().startsWith(queryBase.toLowerCase())) {
        return s;
      }
    }

    // 3.3 Substring match
    for (const s of available) {
      if (s.toLowerCase().includes(queryBase.toLowerCase())) {
        return s;
      }
    }

    // 3.4 Reverse substring match
    for (const s of available) {
      if (queryBase.toLowerCase().includes(s.toLowerCase())) {
        return s;
      }
    }

    return null;
  }
}
